### PR TITLE
[SCEV] Add predicated version of getSymbolicMaxBackedgeTakenCount.

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds a predicated version of getSymbolicMaxBackedgeTakenCount.

The intended use for this is loop access analysis for loops with uncountable exits. When analyzing dependences and computing runtime checks, we need the smallest upper bound on the number of iterations. In terms of memory safety, it shouldn't matter if any uncomputable exits leave the loop, as long as we prove that there are no dependences given the minimum of the countable exits. The same should apply also for generating runtime checks.

### CODE REVIEW

**Comment by fhahn** [comment]
marking as a draft as this isn't ready yet

**Comment by fhahn** [comment]
The PR contains 3 commits, 2 of which can be split off independently 1) refactoring getSymbolicMax to work on BackedgeTakenInfo directly and 2) adding tests for predicated symbolic max.

File: llvm/include/llvm/Analysis/ScalarEvolution.h [hunk 1]
```diff
@@ -912,6 +912,9 @@ class ScalarEvolution {
     return getBackedgeTakenCount(L, SymbolicMaximum);
   }
 
+  const SCEV *getPredicatedSymbolicMaxBackedgeTakenCount(
```
Comment (nikic) [inline llvm/include/llvm/Analysis/ScalarEvolution.h:915]: Doc comment.

File: llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll [hunk 2]
```diff
@@ -0,0 +1,119 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -passes='print<scalar-evolution>' %s -disable-output 2>&1 | FileCheck %s
```
Comment (nikic) [inline llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll:2]: Use `-scalar-evolution-classify-expressions=0` to only print the exit counts.

File: llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
```
Comment (nikic) [inline llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll:4]: Needed?

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll (hunk 3 above)
Comment (fhahn) [reply llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll]: Nope, removed, thanks! Landed the adjusted tests separately in ed4227aad37f2c4adf307b63050fb9aee52b07f8

File: llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll (hunk 2 above)
Comment (fhahn) [reply llvm/test/Analysis/ScalarEvolution/predicated-symbolic-max-backedge-taken-count.ll]: Ah thanks, added in the commit that landed the tests ( ed4227aad37f2c4adf307b63050fb9aee52b07f8 )

File: llvm/include/llvm/Analysis/ScalarEvolution.h (hunk 1 above)
Comment (fhahn) [reply llvm/include/llvm/Analysis/ScalarEvolution.h:919]: Added, thanks!

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 4]
```diff
@@ -8644,9 +8649,9 @@ ScalarEvolution::BackedgeTakenInfo::getConstantMax(ScalarEvolution *SE) const {
   return getConstantMax();
 }
 
-const SCEV *
-ScalarEvolution::BackedgeTakenInfo::getSymbolicMax(const Loop *L,
-                                                   ScalarEvolution *SE) {
+const SCEV *ScalarEvolution::BackedgeTakenInfo::getSymbolicMax(
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Analysis/ScalarEvolution.cpp:8652]: You are right, this should have been referenced, apologies for that! To rectify that, how about reverting the patch and then recommitting with an adjusted message to include an attribution like

```
The extension of getSymbolicMax to support predication is inspired by David Sherwood's (@david-arm) version in https://github.com/llvm/llvm-project/pull/88385
```

Happy to adjust as needed, if you think a different wording would be better

[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Analysis/ScalarEvolution.cpp:8652]: It's ok, there's no need to revert the patch and no harm done! Perhaps just something to remember in future that's all. Thanks!

**Pull request merged**